sdk/mobile-center-push/src/main/java/com/microsoft/azure/mobile/push/Push.java

```diff
@@ -0,0 +1,195 @@
+
+    /**
+     * Constant marking event of the push group.
+     */
+    private static final String PUSH_GROUP = "group_push";
+
+    /**
+     * Base key for stored preferences.
+     */
+    private static final String PREFERENCE_PREFIX = SERVICE_NAME + ".";
+
+    /**
+     * Preference key to store push token.
+     */
+    @VisibleForTesting
+    static final String PREFERENCE_KEY_PUSH_TOKEN = PREFERENCE_PREFIX + "push_token";
+
+    /**
+     * Shared instance.
+     */
+    @SuppressLint("StaticFieldLeak")
+    private static Push sInstance;
+
+    /**
+     * The PNS handle for this installation.
+     */
+    private String mPushToken;
+
+    /**
+     * Remember if we already sent push installation log
+     */
+    private boolean mPushTokenSent;
+
+    /**
+     * Log factories managed by this service.
+     */
+    private final Map<String, LogFactory> mFactories;
+
+    /**
+     * Init.
+     */
+    private Push() {
+        mFactories = new HashMap<>();
+        mFactories.put(PushInstallationLog.TYPE, new PushInstallationLogFactory());
+    }
+
+    /**
+     * Get shared instance.
+     *
+     * @return shared instance.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static synchronized Push getInstance() {
+        if (sInstance == null) {
+            sInstance = new Push();
+        }
+        return sInstance;
+    }
+
+    @VisibleForTesting
+    static synchronized void unsetInstance() {
+        sInstance = null;
+    }
+
+    /**
+     * Check whether Push service is enabled or not.
+     *
+     * @return <code>true</code> if enabled, <code>false</code> otherwise.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static boolean isEnabled() {
+        return getInstance().isInstanceEnabled();
+    }
+
+    /**
+     * Enable or disable Push service.
+     *
+     * @param enabled <code>true</code> to enable, <code>false</code> to disable.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static void setEnabled(boolean enabled) {
+        getInstance().setInstanceEnabled(enabled);
+    }
+
+    /**
+     * Enqueue a push installation log.
+     *
+     * @param pushToken the push token value
+     */
+    private void enqueuePushInstallationLog(@NonNull String pushToken) {
+        if (isInactive())
+            return;
+        if (mPushTokenSent)
+            return;
+        PushInstallationLog log = new PushInstallationLog();
+        log.setPushToken(pushToken);
+        mChannel.enqueue(log, PUSH_GROUP);
+        mPushTokenSent = true;
+    }
+
+    /**
+     * Handle push token update success.
+     *
+     * @param token the push token value
+     */
+    @VisibleForTesting
+    synchronized void onTokenRefresh(@NonNull String token) {
+        mPushToken = token;
+        MobileCenterLog.debug(LOG_TAG, "Push token: " + mPushToken);
+        PreferencesStorage.putString(PREFERENCE_KEY_PUSH_TOKEN, mPushToken);
+        enqueuePushInstallationLog(mPushToken);
```

Comment:
If the token is refreshed but we already sent one during process lifetime we will skip sending which is incorrect behavior. It's not unusual that Google can replace the token with a "canonical" version so we should be ready to accept a new token any time. I would replace the boolean check by a string comparison with previous token value.